lib/puppet/network/formats.rb

```diff
@@ -111,6 +113,31 @@ def data_to_instance(klass, data)
   end
 end
 
+Puppet::Network::FormatHandler.create_serialized_formats(:json, :mime => 'application/json', :weight => 15, :required_methods => [:render_method, :intern_method], :intern_method => :from_data_hash) do
+  def intern(klass, text)
+    data_to_instance(klass, JSON.parse(text))
+  end
+
+  def intern_multiple(klass, text)
+    JSON.parse(text).collect do |data|
+      data_to_instance(klass, data)
+    end
+  end
+
+  # JSON monkey-patches Array, so this works.
+  # https://github.com/ruby/ruby/blob/ruby_1_9_3/ext/json/generator/generator.c#L1416
+  def render_multiple(instances)
+    instances.to_json
+  end
+
+  # Unlike PSON, we do not need to unwrap the data envelope, because legacy 3.x agents
+  # have never supported JSON
```

Comment:
If anyone else is curious what this pson envelope looks like, here's an example: https://github.com/puppetlabs/puppet/blob/3.8.7/lib/puppet/resource/catalog.rb#L388-L396